### ISSUE TITLE
How I can mock response of Apollo 

### ISSUE THREAD

**Issue by Abdelsattar** [apollographql/apollo-android#3300]
Hi 

I'm trying to mock response to my query but I can't because the builder needed and doesn't know how to pass operations?

`  val expectedResponse = Resource.success<Response<MyQuery.Data>>(Response("Response.builder(Operation<>)"))`

I want to know How to build the inside the quote "Response.builder(Operation<>)"?

**Comment by martinbonnin** [comment]
Hi! Can you ellaborate a bit more? I'm not sure what you're expecting in the success case. Since `D` is a generic type parameter and all different queries return data of different type, there's not a single `data` that will work for all cases. Unless maybe `null` but I doubt this is what you're looking for...

**Comment by Abdelsattar** [comment]
I'm expecting data of type MyQuery.Data so I need to build response that contains Mock  MyQuery.Data (Data is apollo generated class of MyQuery response

**Comment by martinbonnin** [comment]
My point is that each different operation will have a different `Data` so you can't instantiate the generic type parameter `D` from your method:

```kotlin
suspend fun <D : Operation.Data, T, V : Operation.Variables> suspendedQuery(
    query: Query<D, T, V>,
    cachePolicy: HttpCachePolicy.Policy = HttpCachePolicy.NETWORK_FIRST
): Resource<Response<T>> {
    val data: D = // You can't create an instance of D here
    return Resource.success(Response.builder().data(data).build())
}
```

You'll either have to create specialized versions of your test queries:

```kotlin
suspend fun suspendedQuery(
    query: Query<MyQuery.Data, MyQuery.Data, MyQuery.Variables>,
    cachePolicy: HttpCachePolicy.Policy = HttpCachePolicy.NETWORK_FIRST
): Resource<Response<T>> {
    val data = MyQuery.Data(...) // It's posible to instanciate MyQuery.Data here
    return Resource.success(Response.builder().data(data).build())
}
```

Or pass a factory method:

```kotlin
suspend fun <D : Operation.Data, T, V : Operation.Variables> suspendedQuery(
    query: Query<D, T, V>,
    cachePolicy: HttpCachePolicy.Policy = HttpCachePolicy.NETWORK_FIRST,
    dataFactory: () -> D
): Resource<Response<T>> {
    val data = dataFactory()
    return Resource.success(Response.builder().data(data).build())
}
```

**Comment by Abdelsattar** [comment]
hi @martinbonnin  I tried the following 

` suspend fun suspendedQuery(
             query: Query< MyQuery.Data, MyQuery.Data, Operation.Variables>,
             cachePolicy: HttpCachePolicy.Policy = HttpCachePolicy.NETWORK_FIRST
          ): Resource<Response< MyQuery.Data>> {
          val data = MyQuery.Data // It's posible to instanciate MyQuery.Data here
         return Resource.success(Response.builder().data(data).build())
      } 
`

But I got the following error 

<img width="707" alt="Screenshot 2021-08-25 at 12 30 28" src="https://user-images.githubusercontent.com/9407394/130774887-8a2ef6e5-5d40-4ab6-af50-71d7d4538347.png">

**Comment by martinbonnin** [comment]
Yup, my bad. You can disambiguate the type parameter with something like below:

```kotlin
suspend fun suspendedQuery(
    query: Query<MyQuery.Data, MyQuery.Data, MyQuery.Variables>,
    cachePolicy: HttpCachePolicy.Policy = HttpCachePolicy.NETWORK_FIRST
): Resource<Response<MyQuery.Data>> {
    val data = MyQuery.Data(...) // It's posible to instanciate MyQuery.Data here
    return Resource.success(Response.builder<MyQuery.Data>().data(data).build())
}
```
